File: days/src/main/scala/aoc2024s/day7/Day7.scala

package aoc2024s.day7;

import utils.IO

import scala.annotation.tailrec
import scala.jdk.CollectionConverters.*

object Day7 {

  enum Operation {
    case Value(value: Long)
    case Add, Multiply, Concatenate
  }

  import Operation.*

  case class Equation(target: Long, numbers: List[Operation])

  object Equation {

    def apply(line: String): Equation = {
      val parts = line.split(": ");
      val target = parts(0).toLong;
      val numbers = parts(1).split(" ").map(n => Value(n.toLong)).toList;
      Equation(target, numbers);
    }
  }

  extension [A](list: List[A])
    def sequences(size: Int): LazyList[List[A]] = {
      if (size == 0) LazyList(List.empty)
      else {
        val rest = sequences(size - 1)
        list.to(LazyList).flatMap { x => rest.map(x :: _) }
      }
    }
    def zipAppend(other: List[A]): List[A] = {
      assert(list.size == other.size)
      list.zip(other).flatMap { case (a, b) => List(a, b) }
    }

  object Combinator {

    def combine(numbers: List[Operation], ops: LazyList[List[Operation]]): LazyList[List[Operation]] = {
      ops match {
        case LazyList() => LazyList()
        case head #:: tail => {
          val combined = numbers.head :: head.zip(numbers.tail).flatMap { case (op, num) => List(op, num) }
          combined #:: combine(numbers, tail)
        }
      }
    }

    def canBeTrue1(equation: Equation): Boolean = {
      val intermediateOps = List(Add, Multiply).sequences(equation.numbers.size - 1)
      val possibleExpressions = combine(equation.numbers, intermediateOps)
      possibleExpressions.exists { expression =>
        evaluatesTo(expression, equation.target)
      }
    }

    def canBeTrue2(equation: Equation): Boolean = {
      // if (canBeTrue1(equation)) return true
      val intermediateOps = List(Add, Multiply, Concatenate).sequences(equation.numbers.size - 1)
      val possibleExpressions = combine(equation.numbers, intermediateOps)
      possibleExpressions.exists { expression =>
        evaluatesTo(expression, equation.target)
      }
    }

    def evaluatesTo(expression: List[Operation], target: Long): Boolean = {
      @tailrec
      def go(acc: Long, ops: List[Operation], numbers: List[Operation]): Boolean = {
        if (acc > target) false
        else
          numbers match {
            case Nil => acc == target
            case Add :: Value(value) :: tail => go(acc + value, ops, tail)
            case Multiply :: Value(value) :: tail => go(acc * value, ops, tail)
            case Concatenate :: Value(value) :: tail => go((acc.toString ++ value.toString).toLong, ops, tail)
            case _ => sys.error("Invalid expression")
          }
      }

      expression.head match {
        case Value(value) => go(value, Nil, expression.tail)
        case _ => sys.error("Invalid expression")
      }
    }
  }

  def part1(data: List[String]): Long = {
    data
      .map(Equation(_))
      .filter(Combinator.canBeTrue1)
      .map(_.target)
      .sum
  }

  def part2(data: List[String]): Long = {
    data
      .map(Equation(_))
      .filter(Combinator.canBeTrue2)
      .map(_.target)
      .sum
  }

  @main def main7(): Unit = {
    val data = IO.getResourceAsList("aoc2024/day7.txt").asScala.toList;
    val part1 = Day7.part1(data);
    println(s"part1 = $part1");
    val part2 = Day7.part2(data);
    println(s"part2 = $part2");
  }
}

File: days/src/main/java/aoc2024j/day7/Day7.java

package aoc2024j.day7;

import utils.IO;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public class Day7 {

    enum Operation {
        Add, Multiply, Concatenate
    }

    record Equation(long target, ArrayList<Long> numbers) {
        static Equation parse(String line) {
            var parts = line.split(": ");
            var target = Long.parseLong(parts[0]);
            var numbers = Arrays.stream(parts[1].split(" "))
                    .map(Long::parseLong)
                    .collect(Collectors.toCollection(ArrayList::new));
            return new Equation(target, numbers);
        }
    }

    record Node<A>(A value, Node<A> next) {
    }

    static class CachedCombinator<A> {
        final ArrayList<List<Node<A>>> cache = new ArrayList<>();
        final List<A> list;

        CachedCombinator(List<A> list) {
            this.list = list;
        }

        List<Node<A>> get(int n) {
            if (n < cache.size()) {
                return cache.get(n);
            } else if (n == 0) {
                var result = new ArrayList<Node<A>>(1);
                result.add(null);
                cache.add(result);
                return result;
            } else {
                var result = get(n - 1).stream()
                        .flatMap(parent -> list.stream().map(head -> new Node<>(head, parent)))
                        .toList();
                cache.add(result);
                return result;
            }
        }
    }

    static class Generator {

        static long concatLong(long a, long b) {
            long acc = a * 10;
            long n = b / 10;
            while (n > 0) {
                acc *= 10;
                n /= 10;
            }
            return acc + b;
        }

        static boolean generateAndTest(long target, ArrayList<Long> numbers, Node<Operation> ops) {
            int i = 0;
            var acc = numbers.get(i++);
            while (acc <= target && ops != null) {
                var n = numbers.get(i++);
                var op = ops.value();
                ops = ops.next();
                switch (op) {
                    case Add -> acc += n;
                    case Multiply -> acc *= n;
                    case Concatenate -> acc = concatLong(acc, n);
                }
                if (acc > target) {
                    return false;
                }
            }
            return acc == target;
        }

        static boolean canBeTrue(CachedCombinator<Operation> cachedCombinator, Equation equation) {
            var ops = cachedCombinator.get(equation.numbers.size() - 1);
            return ops.stream()
                    .anyMatch(op ->
                            generateAndTest(equation.target, equation.numbers, op));
        }
    }

    long part1(List<String> data) {
        var combinator = new CachedCombinator<>(List.of(Operation.Add, Operation.Multiply));
        return data.stream()
                .map(Equation::parse)
                .filter(equation -> Generator.canBeTrue(combinator, equation))
                .mapToLong(Equation::target)
                .sum();
    }

    long part2(List<String> data) {
        var combinator = new CachedCombinator<>(List.of(Operation.Add, Operation.Multiply,
                Operation.Concatenate));
        return data.stream()
                .map(Equation::parse)
                .filter(equation -> Generator.canBeTrue(combinator, equation))
                .mapToLong(Equation::target)
                .sum();
    }

    public static void main() {
        var day7 = new Day7();
        var data = IO.getResourceAsList("aoc2024/day7.txt");
        var part1 = day7.part1(data);
        System.out.println("part1 = " + part1);
        var part2 = day7.part2(data);
        System.out.println("part2 = " + part2);
    }
}